python/lsst/daf/butler/_butler.py

```diff
@@ -1597,6 +1613,142 @@ def retrieveArtifacts(
             overwrite=overwrite,
         )
 
+    def exists(
+        self,
+        datasetRefOrType: DatasetRef | DatasetType | str,
+        /,
+        dataId: DataId | None = None,
+        *,
+        full_check: bool = True,
+        collections: Any = None,
+        **kwargs: Any,
+    ) -> DatasetExistence:
+        """Indicate whether a dataset is known to Butler registry and
+        datastore.
+
+        Parameters
+        ----------
+        datasetRefOrType : `DatasetRef`, `DatasetType`, or `str`
+            When `DatasetRef` the `dataId` should be `None`.
+            Otherwise the `DatasetType` or name thereof.
+            If a resolved `DatasetRef`, the associated dataset
+            is returned directly without additional querying.
+        dataId : `dict` or `DataCoordinate`
+            A `dict` of `Dimension` link name, value pairs that label the
+            `DatasetRef` within a Collection. When `None`, a `DatasetRef`
+            should be provided as the first argument.
+        full_check : `bool`, optional
+            If `True`, an additional check will be made for dataset artifact
+            existence. This will involve additional overhead due to the need
+            to query an external system. If `False` registry and datastore
+            will solely be asked if they know about the dataset but no
+            check for the artifact will be performed.
+        collections : Any, optional
+            Collections to be searched, overriding ``self.collections``.
+            Can be any of the types supported by the ``collections`` argument
+            to butler construction.
+        **kwargs
+            Additional keyword arguments used to augment or construct a
+            `DataCoordinate`.  See `DataCoordinate.standardize`
+            parameters.
+
+        Returns
+        -------
+        existence : `DatasetExistence`
+            Object indicating whether the dataset is known to registry and
+            datastore. Evaluates to `True` if the dataset is present is known
+            to both.
+        """
+        existence = DatasetExistence.Unknown
+
+        if isinstance(datasetRefOrType, DatasetRef):
+            registry_ref = self.registry.getDataset(datasetRefOrType.id)
+            if registry_ref is not None:
```

Comment:
I was not expecting a situation where a UUID could match but somehow be associated with a completely different ref. Doesn't that seem unlikely? I can add the check of dataId and datasetType if you think it's possible. I really need a `DatasetRef.is_compatible_with` call or something (since `__eq__` does not check for storage class compatibility)